chain/consensus/actors/atomic-replace/replace.go

```diff
@@ -0,0 +1,264 @@
+	// Anyone can lock the state
+	rt.ValidateImmediateCallerAcceptAny()
+
+	var st ReplaceState
+	rt.StateTransaction(&st, func() {
+		switch params.Method {
+		case 5:
+			builtin.RequireNoErr(rt, st.Owners.LockState(), exitcode.ErrIllegalArgument, "error locking state")
+		default:
+			rt.Abortf(exitcode.ErrIllegalArgument, "provided method doesn't support atomic execution. No need to lock")
+		}
+	})
+
+	c, err := st.Owners.Cid()
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalArgument, "error computing Cid for locked state")
+	return &atomic.LockedOutput{Cid: c}
+}
+
+func (st *ReplaceState) unlock(rt runtime.Runtime) {
+	builtin.RequireNoErr(rt, st.Owners.UnlockState(), exitcode.ErrIllegalArgument, "error unlocking state")
+}
+
+func (a ReplaceActor) Merge(rt runtime.Runtime, params *atomic.MergeParams) *abi.EmptyValue {
+	// Only system actor can trigger this function.
+	rt.ValidateImmediateCallerIs(builtin.SystemActorAddr)
+	var st ReplaceState
+	rt.StateTransaction(&st, func() {
+		merge := &Owners{}
+		err := atomic.UnwrapMergeParams(params, merge)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error unwrapping output from mergeParams")
+		st.merge(rt, merge)
+	})
+
+	return nil
+}
+
+func (a ReplaceActor) Unlock(rt runtime.Runtime, params *atomic.UnlockParams) *abi.EmptyValue {
+	// Only system actor can trigger this function.
+	rt.ValidateImmediateCallerIs(builtin.SystemActorAddr)
+
+	var st ReplaceState
+	rt.StateTransaction(&st, func() {
+		output := &Owners{}
+		err := atomic.UnwrapUnlockParams(params, output)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error unwrapping output from unlockParams")
+		switch params.Params.Method {
+		case 5:
+			st.merge(rt, output)
+			st.unlock(rt)
+		default:
+			rt.Abortf(exitcode.ErrIllegalArgument, "this method has nothing to merge")
+		}
+	})
+
+	return nil
+}
+
+func (st *ReplaceState) merge(rt runtime.Runtime, state atomic.LockableState) {
+	owners := &Owners{}
+	err := atomic.UnwrapLockableState(st.Owners, owners)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "error unwrapping owners")
+	builtin.RequireNoErr(rt, owners.Merge(state), exitcode.ErrIllegalState, "error merging output")
+	st.storeOwners(rt, owners)
+}
+
+func (a ReplaceActor) Abort(rt runtime.Runtime, params *atomic.LockParams) *abi.EmptyValue {
+	// FIXME: We should check here that the only one allowed to abort an execuetion is
+	// the rt.Caller() that locked the state? Or is the system.actor because is triggered
+	// through a top-down transaction?
+	rt.ValidateImmediateCallerAcceptAny()
+
+	var st ReplaceState
+	rt.StateTransaction(&st, func() {
+		switch params.Method {
+		case 5:
+			st.unlock(rt)
+		default:
+			rt.Abortf(exitcode.ErrIllegalArgument, "this method has nothing to unlock")
+		}
+	})
+
+	return nil
+}
+
+func ValidateLockedState(rt runtime.Runtime, st *ReplaceState) {
+	builtin.RequireNoErr(rt,
+		atomic.ValidateIfLocked([]*atomic.LockedState{st.Owners}...),
```

Comment:
Could st be nil?

Reply:
Similar to `params`, `st` can't be `nil` as it is the result of unmarshalling the actor state from the stateStore. Actor state is initialized in actors' constructor, so having this would mean that something very wrong happened.